src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.cs

```diff
@@ -66,92 +65,165 @@ public async Task FindReferencesAsync(ISymbol symbol, CancellationToken cancella
+
+                // Report the initial set of symbols to the caller.
+                var allSymbols = symbolSet.GetAllSymbols();
+                await ReportGroupsAsync(allSymbols, cancellationToken).ConfigureAwait(false);
+
+                // Determine the set of projects we actually have to walk to find results in.  If the caller provided a
+                // set of documents to search, we only bother with those.
+                var projectsToSearch = await GetProjectIdsToSearchAsync(symbolSet.GetAllSymbols(), cancellationToken).ConfigureAwait(false);
 
-                var projectMap = await CreateProjectMapAsync(symbols, cancellationToken).ConfigureAwait(false);
-                var projectToDocumentMap = await CreateProjectToDocumentMapAsync(projectMap, cancellationToken).ConfigureAwait(false);
-                ValidateProjectToDocumentMap(projectToDocumentMap);
+                // We need to process projects in order when updating our symbol set.  Say we have three projects (A, B
+                // and C), we cannot necessarily find inherited symbols in C until we have searched B.  Importantly,
+                // while we're processing each project linearly to update the symbol set we're searching for, we still
+                // then process the projects in parallel once we know the set of symbols we're searching for in that
+                // project.
+                var dependencyGraph = _solution.GetProjectDependencyGraph();
+                await _progressTracker.AddItemsAsync(projectsToSearch.Count, cancellationToken).ConfigureAwait(false);
+
+                using var _1 = ArrayBuilder<Task>.GetInstance(out var projectTasks);
+
+                foreach (var projectId in dependencyGraph.GetTopologicallySortedProjects(cancellationToken))
+                {
+                    if (!projectsToSearch.Contains(projectId))
+                        continue;
+
+                    var currentProject = _solution.GetRequiredProject(projectId);
+
+                    // As we walk each project, attempt to grow the search set appropriately up and down the 
+                    // inheritance hierarchy.  Note: this has to happen serially which is why we do it in this
+                    // loop and not inside the concurrent project processing that happens below.
+                    await symbolSet.InheritanceCascadeAsync(currentProject, cancellationToken).ConfigureAwait(false);
+
+                    // Report any new symbols we've cascaded to to our caller.
+                    allSymbols = symbolSet.GetAllSymbols();
+                    await ReportGroupsAsync(allSymbols, cancellationToken).ConfigureAwait(false);
+
+                    projectTasks.Add(Task.Factory.StartNew(
+                        () => ProcessProjectAsync(currentProject, allSymbols, cancellationToken),
+                        cancellationToken, TaskCreationOptions.None, _scheduler).Unwrap());
```

Comment:
Is this the correct pattern for starting a background work with a custom schedule? Will work correctly be cancelled on VS shutdown?

Reply:
1. yes, this is the correct pattern for starting gc work witha  custom scheduler.  THis is already used elsewhere in the findrefs code to allow us to either:
  a. use all cores and compute results fast if the user explicitly asks for a find refs result
  b. use just a single serial scheduler so we're onlyh occupying a core for BG find refs work (for example: for codelens)
2. this work will be stopped if cancellationToken is triggered.  It's up to the caller to do that.  In this case, teh caller could be many different features.  it's up to them to appropriate stop this work based on environmental factors.